src/core_modules/capture-core/utils/cachedDataHooks/useValidatedIDsFromCache.js

```diff
@@ -0,0 +1,75 @@
+// @flow
+import { useCallback, useEffect, useState } from 'react';
+import {
+    containsKeyInStorageAsync,
+    getCachedSingleResourceFromKeyAsync,
+} from '../../MetaDataStoreUtils/MetaDataStoreUtils';
+import { userStores } from '../../storageControllers/stores';
+
+type Props = {|
+    programId: string,
+    orgUnitId?: string,
+|}
+
+export const IdTypes = Object.freeze({
+    PROGRAM_ID: 'programId',
+    ORG_UNIT_ID: 'orgUnitId',
+});
+
+export const useValidatedIDsFromCache = ({ programId, orgUnitId }: Props) => {
+    const [valid, setValid] = useState({
+        [IdTypes.PROGRAM_ID]: undefined,
+        [IdTypes.ORG_UNIT_ID]: undefined,
+    });
+    const [error, setError] = useState();
+
+    const getPromises = useCallback(() => {
+        const promises = [];
+        const keys = [];
+        if (programId) {
+            keys.push({ id: programId, type: IdTypes.PROGRAM_ID, convert: value => value });
+            promises.push(containsKeyInStorageAsync(userStores.PROGRAMS, programId));
+        }
+
+        if (orgUnitId) {
+            keys.push({ id: orgUnitId, type: IdTypes.ORG_UNIT_ID, convert: value => !!value?.organisationUnits[orgUnitId] });
+            promises.push(getCachedSingleResourceFromKeyAsync(userStores.ORGANISATION_UNITS_BY_PROGRAM, programId));
+        }
+
+        Promise.all(promises)
+            .then((adapterResponses) => {
+                // $FlowFixMe - Missing declaration for U
+                const adapterResponseObjects = keys.reduce((acc, { id, type, convert }, index) => {
+                    acc[type] = {
+                        id,
+                        valid: convert(adapterResponses[index]),
```

Comment:
Hey, 
It looks a bit strange to me to call `adapterResponses[index]` with the `index` coming from the reducer of `keys` array. Can we match `keys` and `adapterResponses` using the `id` instead?

One way to do that is to add the `id` (`orgUnitId` and `programId`) as part of the promises response `adapterResponses`. It can then be used like this `adapterResponses.find(adapterResponse.id===id)`

Does that make sense? Thanks!

Reply:
I agree that it's not exactly intuitive. For the programId in this case, it makes sense to perhaps add the key to the `contains`-method in `IndexedDBAdapter`, and then rewrite its usages. For the organization units, it would require a rewrite of the `get`-method. This get-method has different use cases depending on which userStore is passed in. In this case it's the `ORGANISATION_UNITS_BY_PROGRAM`. The method has no knowledge of which orgUnit we want to check for, so the only key passed in is the programId. I'm struggling a bit with finding a solution to how we would implement this for the orgUnitSearch. Do you have any thoughts on this? Do you still think it makes sense with a rewrite?
Thanks!

Reply:
Good morning,
You don't have to rewrite `contains` or `get` from `IndexedDBAdapter`. You can act one call after by modifying the promises responses in `MetaDataStoreUtils.js`. There you can add whatever extra information you need later on. 

I rise a PR https://github.com/dhis2/capture-app/pull/2270 to your branch to showcase one way you can do that. Let me know if it makes sense or you have doubts. 
Thank you!